rails_programming/hotwire/turbo.md

```diff
@@ -0,0 +1,494 @@
+
+#### Turbo Stream Template
+
+You create your Turbo Stream file inside of your `views` folder the same way as any other view. In this example, it would be located at `views/posts/create.turbo_stream.erb`. Our view file will look like this:
+
+~~~erb
+<%= turbo_stream.append "posts", @post %>
+~~~
+
+That line is all we need! What this does is create a Turbo Stream packet with the `append` action. The target of the action is `"posts"`. This is the `<div>` with the `id="posts"` that contains all of our posts. Turbo Stream will locate the `<div>` and append our brand new `@post` to the bottom! Our newly created `@post` will even use the `_post.html.erb` partial to be in the proper layout as all of our other posts. For that, you can thank the Rails naming convention. If you are breaking from Rails convention, you can specify a partial to use inside of your `turbo_stream` template.
+
+And just like that, we can create a post and watch it show up on the list!
+
+#### Additional Turbo Stream Tips
+
+There's no time to get more in-depth on how Turbo Stream works in this lesson, but here are some topics for you to begin your personal research
+if you would like to learn more:
+
+1.  You can chain multiple `turbo_stream` actions in one `turbo_stream.erb` file, like if you wanted to append to one region and
+    update another region at the same time.
+
+2.  You don't *have* to create a separate layout file for your Stream formats. For instance, we could instead change the line in our controller to:
+
+    ~~~ruby
+    format.turbo_stream { render turbo_stream: turbo_stream.append('posts', @post) }
+    ~~~
+
+    However, [you shouldn't do this for anything complex or chained](https://github.com/hotwired/turbo-rails/issues/77#issuecomment-757349251).
+
+3.  You may have noticed that when you submit a new Post, the text box doesn't clear out! That's why Hotwire comes with Stimulus. In the next lesson, we will cover how to create a simple Stimulus controller to reset the new post form.
+
+### Turbo Native
+
+The final piece of Turbo is something that you don't need to know much about for this course, other than it exists. Turbo Native is useful for wrapping your application inside of a mobile-friendly shell. If you are interested in mobile development and how you can use Turbo with it, you may enjoy looking at the demos for [iOS](https://github.com/hotwired/turbo-ios/blob/main/Demo/README.md) and [Android](https://github.com/hotwired/turbo-android/blob/main/demo/README.md) phones.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Single Page Applications
+
+1. Read sections 1, 2, and 3 of Bloomreach's [What Is A Single Page Application](https://www.bloomreach.com/en/blog/2018/07/what-is-a-single-page-application.html#whatssingle-page-application)
+   article
+
+#### Hotwire
+
+1. Watch the [Hotwire Demo Video](https://www.youtube.com/watch?v=eKY-QES1XQQ)
```

Comment:
I don't think his intro video is very good. I only understood a lot of stuff, after hearing via other sources about it and watching it again after a considerable time. I wonder if not an intro video from Gorails or Drifting Ruby would be better:

**Disclaimer: I have not watched these yet**

- https://gorails.com/episodes/hotwire-rails
- https://www.driftingruby.com/episodes/hotwire
- https://www.driftingruby.com/episodes/diving-into-hotwire

Reply:
Update on this: Those 3 videos are all great but not quite what I think this lesson needs. I will try and find a suitable Assignment video because I do agree that a video guide would be useful here. It just takes time to watch through each one as I find it and I wanted to push the other changes that I already had completed.

The first two dive straight into using Turbo Stream as a broadcast from the model over Websocket, which is not the implementation that is being taught here. While that implementation is powerful and is potentially the most common use-case, this lesson does not include that content due to it's placement in the overall curriculum. I think even with a note in the lesson that the implementation is different, it could lead to confusion if used as a reference by a learner who doesn't actually realize the difference exists.

The last video is more technical and has to do with the installation and other 'extra' information, but also does demonstrate Hotwire in use. Im definitely tempted to include it as a bonus resource because the information could be valuable. Putting a pin in that for this moment in time until I can re-watch it later today and at the same time search for a more suitable video for the Assignment section.